### PR TITLE
[Resolves #27] Improve api in class init

### LINKED ISSUE #27: Improve API in class __init__ functions
* enforce kwarg-only arguments to make instance creation more explicit
* ~~remove the `columns` & `colnames` parameters from table-based class inits since we're defining those at declaration rather than runtime~~ (edit: this seems to not be feasible as these parameters are used by the build backend when building a container from disk. So this improvement will not be made.)

### PR DESCRIPTION
**Pull request by dsleiter**
* require keyword-only arguments in __init__ functions to make class instantiation more explicit
* improve __init__ docstrings
* move update_docval utility function to a utils module and improve the docstring to better explain usage

I also tried to remove the need for __init__ params columns and colnames in containers that derive from DynamicTable since those are defined at definition time, however, I discovered those are used when building the container from disk, so they need to remain.

### CODE REVIEW

File: docs/source/credits.rst [hunk 1]
```diff
@@ -8,9 +8,13 @@ AE Studio
...
-Sumner L Norman
-Darin Erat Sleiter
-José Ribeiro
+Sumner L Norman (@sumner15)
+Darin Erat Sleiter (@dsleiter)
+José Ribeiro (@ribeirojose)
+
+Contributors
+============
+Rob Luke (@rob-luke)
```
Comment (rob-luke) [inline docs/source/credits.rst]: ❤️

File: src/pynwb/ndx_nirs/__init__.py [hunk 2]
```diff
@@ -33,27 +33,12 @@
...
-
-
-sources_docval = update_docval(
+_sources_docval = update_docval(
     DynamicTable.__init__,
-    name={"default": "sources"},
-    description={"default": "A table describing optical sources of a NIRS device"},
+    updates=dict(
+        name={"default": "sources"},
+        description={"default": "A table describing optical sources of a NIRS device"},
```
Comment (rob-luke) [inline src/pynwb/ndx_nirs/__init__.py]: ??? 😉 
```suggestion
        description={"default": "A table describing optical sources of a NIRS device."},
```

File: src/pynwb/ndx_nirs/__init__.py [hunk 3]
```diff
@@ -119,15 +124,39 @@ def __init__(self, **kwargs):
...
-    """A DynamicTable representing the optical channels of this NIRS device"""
+    """A DynamicTable representing the optical channels of this NIRS device.
+
+    Before adding channels to the table, the sources and detectors for the
+    device need to be defined in a NIRSSourcesTable and a NIRSDetectorsTable,
+    respectively. These can be passed in either during class initializing or
+    afterwards.
+
+    Examples:
+    ```
```
Comment (rob-luke) [inline src/pynwb/ndx_nirs/__init__.py]: I am not sure what doc generation is being used, but do you need to specify python for highlighting?
Regardless, I love a good example in the docs, very handy.

File: src/pynwb/tests/unit/test_ndx_nirs.py [hunk 4]
```diff
@@ -53,6 +23,7 @@ def test_empty_constructor(self):
 
         self.assertEqual(table.name, "sources")
         self.assertIsInstance(table.description, str)
+        # assert False
```
Comment (rob-luke) [inline src/pynwb/tests/unit/test_ndx_nirs.py]: Mistakenly committed?
```suggestion
```

**Review by rob-luke — APPROVED** [review]
Looks great. Easy to read and review.

File: src/pynwb/ndx_nirs/__init__.py (hunk 3 above)
Comment (dsleiter) [reply src/pynwb/ndx_nirs/__init__.py]: Good point. We don't have doc generation working yet, that's on our backlog, so I'm not sure yet. The extension template included some sphinx integration, so I imagine we'd go with that, but we haven't gotten it working yet (or decided on a docstring style). I'll mark the block as python for now, and we'll need to check all docstrings once we get the doc generation working.

File: src/pynwb/tests/unit/test_ndx_nirs.py (hunk 4 above)
Comment (dsleiter) [reply src/pynwb/tests/unit/test_ndx_nirs.py]: good catch, thanks!

File: src/pynwb/ndx_nirs/__init__.py (hunk 2 above)
Comment (dsleiter) [reply src/pynwb/ndx_nirs/__init__.py]: this is fixed in #30 :-)

**Comment by dsleiter** [comment]
Awesome, thanks for the review @rob-luke!

**Pull request merged**